### DISCUSSION TITLE
Plan for Jolokia 2.0

### DISCUSSION THREAD

**Discussion by tadayosi** [rhuss/jolokia#555] [General]
We will shortly have an update from @rhuss on the transfer of Jolokia project from his own repo to organisation. It is to keep Jolokia up to date by the community.

One of the major goals of the community-based Jolokia development would be releasing Jolokia 2.0. We internally discussed the ideas about Jolokia 2.0 in advance, and here are the summary of the discussion:

- Keep Jolokia 1.x for backward compatibilies; keeping support for the old Java versions including 6,7,8.
- Migrate to Jakarta EE with Jolokia 2.0

(@rhuss @grgrzybek @apupier Please correct me if I'm misunderstanding it.)

Here I opened a GitHub discussion so that this topic is shared with the community and can be discussed further with wider audience. Please share your thoughts. Thank you!

**Comment by tadayosi** [comment]
@skarsaune Please join the discussion!

**Comment by grgrzybek** [comment]
> * Keep Jolokia 1.x for backward compatibilies; keeping support for the old Java versions including 6,7,8.
> * Migrate to Jakarta EE with Jolokia 2.0

That's exactly the plan. Recently I brought back real JDK6 compatibility (ff4f020a994ac94944d0caf4e5d39cb424e50eb8) while keeping all the tests working on JDK17.
From Jakarta EE 9+ (effectively - Jakarta EE 10) point of view, there are few things to do, but for now I focus on the OSGi side of things outside of Jolokia.

**Comment by dennisgerhardt** [comment]
I recently switched to springboot3 and unfortunately jolokia doesn't seem to work there anymore. All in all I find the library (jolokia) perfect to allow an admin view on the application

**Comment by grgrzybek** [comment]
See also #568.

**Comment by perspilling** [comment]
Hi. Do you have an idea when Jolokia 2 with support for Spring Boot 3 will be available?

**Comment by grgrzybek** [comment]
We're in the process of preparing a release, which should happen within few next weeks. Spring Boot 3 related stuff may be added/checked before or after the release though.

So please give us some time.

**Reply by perspilling** [reply]
Ok, very good. Looking forward to it.

**Reply by tadayosi** [reply]
@perspilling In the meantime, we've already released a milestone 2.0.0-M4. You can try it early and see if it works as you expect.

**Reply by perspilling** [reply]
I can try that. What Maven dependency should I use in the client application?

**Reply by perspilling** [reply]
In the SB 2 app I use jolokia-core

**Comment by grgrzybek** [comment]
> In the SB 2 app I use jolokia-core

Yes, because artifacts in Jolokia 2.0 were reorganized back in 2016 ;) Part of Jolokia 2 documentation will be a listing of all available Maven artifacts. Please for now see the description of #557.

**Comment by grgrzybek** [comment]
Please start with 3 jars:
* org.jolokia:jolokia-server-core
* org.jolokia:jolokia-service-jmx
* org.jolokia:jolokia-service-serializer

**Reply by perspilling** [reply]
I added this to a vanilla hello-world SB 3 app with spring-boot-admin-starter-client 3.1.5 (generated by https://start.spring.io). It didn't work. It seems that the JmxServiceActivator is not called.

**Comment by grgrzybek** [comment]
> I added this to a vanilla hello-world SB 3 app with spring-boot-admin-starter-client 3.1.5 (generated by https://start.spring.io). It didn't work. It seems that the JmxServiceActivator is not called.

Can you share your project set up via github?

**Reply by perspilling** [reply]
Yes, here: https://github.com/perspilling/sba-example

**Reply by grgrzybek** [reply]
Thanks, let me check.

**Comment by grgrzybek** [comment]
Please see perspilling/sba-example#1 where I've added some changes to your sba-example/sba-server project and it should work.

**Comment by mbazos** [comment]
Hi, I am looking to use Jolokia in my projects but ran into a problem. What I am most interested in is using Jolokia to do JFR recordings from JMC. I did the setup using a Java Agent approach and installed the plugin from @skarsaune 
but when I attempt to do a JFR recording I run into an exception. I tried using Jolokia 1.7.3 and the new 2.0 Milestone but it failed for the same reason. Please see here:
https://github.com/skarsaune/jmc-cloud-extensions/issues/3#issue-1983992206

Sorry for cross posting this but really wanted to know is there a way to get a JFR recording snapshot using Jolokia? I thought using jmc-cloud-extensions would do the trick but I get an exception when attempting to run it. Any thoughts or suggestions would be helpful.

Also I did this following the recommendations from the docs:

<img width="1067" alt="Screenshot 2023-11-08 at 12 38 00 PM" src="https://github.com/jolokia/jolokia/assets/1580583/ce857219-e3e7-4277-9cab-023a88b6156e">

There was a recording that @skarsaune did which shows doing the JFR recording but not sure of all the right versions of things to make this possible.

**Comment by grgrzybek** [comment]
Thanks for the report. There must be some problem if it doesn't work with 1.7.3.
I never worked with JFR and I'm glad I finally have the opportunity ;).

**Reply by mbazos** [reply]
Thanks for the quick reply, I am not sure if it's an issue with 1.7.3 as I did try the 2.0 version and it failed as well with the same exception. I am just not familiar enough to know if the error is on the jolokia side sending some data which the client isn't expecting or something with the plugin which was developed.

It might not be a terrible idea to contact @skarsaune to see if the plugin should be part of the Jolokia project as I imagine there could be some compatibility issues there as new versions of Jolokia are published.

JFR is awesome it's the ultimate profiler for Java and being able to adhoc take snapshots over http is very interesting.

**Reply by mbazos** [reply]
@grgrzybek I am going to attempt to do this Jolokia 1.7.0 that seems to be the same version of the Jolokia that the plugin by @skarsaune uses.

**Reply by mbazos** [reply]
Still seems to be the same issue with 1.7.0 :(

```
Caused by: java.lang.ClassCastException: class org.json.simple.JSONArray cannot be cast to class [Ljavax.management.openmbean.CompositeData; (org.json.simple.JSONArray is in unnamed module of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @57b52b8a; [Ljavax.management.openmbean.CompositeData; is in module java.management of loader 'bootstrap')
	at org.openjdk.jmc.rjmx.services.jfr.internal.FlightRecorderServiceV2.getAvailableRecordings(FlightRecorderServiceV2.java:297)
	at org.openjdk.jmc.flightrecorder.controlpanel.ui.wizards.RecordingWizardModel.getDescriptors(RecordingWizardModel.java:320)
	at org.openjdk.jmc.flightrecorder.controlpanel.ui.wizards.RecordingWizardModel.findUniqueName(RecordingWizardModel.java:287)
	at org.openjdk.jmc.flightrecorder.controlpanel.ui.wizards.RecordingWizardModel.<init>(RecordingWizardModel.java:164)
	at org.openjdk.jmc.flightrecorder.controlpanel.ui.actions.StartRecordingAction.doCreateWizard(StartRecordingAction.java:68)
	at org.openjdk.jmc.ui.wizards.AbstractWizardUserAction.createWizard(AbstractWizardUserAction.java:65)
	at org.openjdk.jmc.ui.wizards.WizardAdapterFactory.getAdapter(WizardAdapterFactory.java:45)
	... 76 more	
```

**Comment by skarsaune** [comment]
I can try to reproduce. When it comes to reverse mapping of types, support
may have to be gradually added in some cases, to support new use cases.
Will try to look at both permanent solutions in jolokia, and whether there
are any workarounds (manually overriding types from the client side)

**Reply by mbazos** [reply]
Thanks @skarsaune I haven't directly worked on the jolokia library but if there is anything I can do to help please let me know. Appreciate the prompt reply on this.

**Reply by mbazos** [reply]
@grgrzybek Just wanted to report back. So I did get JFR recording working with Jolokia and a custom branch of JMC that @skarsaune made. Please see here if you want to replicate it https://github.com/skarsaune/jmc-cloud-extensions/issues/3#issuecomment-1808383557

I didn't review the changes @skarsaune made to JMC to get it to work but the version of JMC is pretty out of date and as mentioned would be great if this could work with the latest off the shelf JMC instead of a custom build.


![Screenshot 2023-11-13 at 10 31 27 AM](https://github.com/jolokia/jolokia/assets/1580583/664dc988-d1e2-4675-bfbf-c02a020e1f66)

**Reply by skarsaune** [reply]
Absolutely agree with all your points @mbazos . I aim to both update the plugins and first and foremost work on the PR to get it merged to jmc proper. With regards to the specific issue accessing fligth recordings over jolokia, I created a separate issue to track progress on troubleshooting of that: https://github.com/jolokia/jolokia/issues/613
